### ISSUE TITLE
Add Functor Support To ChaiScript

### ISSUE THREAD

**Issue by stephenberry** [ChaiScript/ChaiScript#312]
I would like to be able to call a functor within ChaiScript.

Example C++:
``` C++
struct Functor
{
   void operator()() { std::cout << 5 << '\n'; }
};
```

Registration:
``` C++
chai.add(chaiscript::user_type<Functor>(), "Functor");
chai.add(chaiscript::constructor<Functor()>(), "Functor");
chai.add(chaiscript::fun([](Functor& functor) { functor(); }), "()");
```

Ideally this would allow me to perform the following in script:
```
var functor = Functor()
functor() // should print "5"
```

Currently ChaiScript throws: `Error: "'functor' does not evaluate to a function."

**Comment by lefticus** [comment]
I think what would need to happen would be to allow for registering `operator()` overloads, as you are suggesting, but that doesn't currently exist.

I can see the appeal, but think this is low priority. largely because I'm not sure about the potential performance implications of needing **yet one more way** of trying to execute a function.

**Comment by stephenberry** [comment]
I agree with keeping this a low priority if you think it may affect performance. I just thought I'd pose the idea because it helps with keeping ChaiScript matching C++ code. It's easy enough to name my functor `operator()` calls, but it loses similarity with my C++ code.

**Comment by Royal-806** [comment]
I'm still learning some of the more complex topics so forgive me if this is completely incorrect.
If adding more ways to execute a function adds potential performances implications could you not just make one way of executing the function that is highly optimized. Then with that one best way build other methods on top of it? Simply leaving convenience/performance choices to the user?

**Comment by stephenberry** [comment]
I believe lefticus is referring to parsing speed. The actual calls to the functions will not be slowed by adding in this support, but the parser needs to determine how the function is being called. Adding another means of executing a function call adds complexity to the parser, which in turn means the run time performance of executing the script could be slowed.

**Comment by Royal-806** [comment]
Oh wow I can't believe I didn't think of the parser. That makes total sense.

**Comment by lefticus** [comment]
I actually did mean a runtime thing. If I see a function call operation, I'd have to check: "is this a function? If not, does it have an `operator()`?", then I'd have to look up the `operator()` and dispatch *that* as a function call, passing the remaining arguments to it.

I currently do this for functions that happen to be parameters of an object.

Example:

```
obj.property = fun(){};
obj.property();
```

This has to look up `property` see if it's a function name (since properties/member variables are also functions), execute it to get back the function object that function returns, then call the returned function object.

This gets more interesting if the function object takes any parameters, because it might need to have the implicit `this` object passed to it.

So, while I like the idea of full functor support, I'm reluctant to layer an extra level of complexity on this process.

So, for example you did this:

```
obj.property = some_custom_function_object;
obj.property();
```

This would be:

* Look up "property"
* Execute "property"
* See if the return value is a function object
* It's not, see if we want to call it via a custom `operator()` that's been registered

OTOH, a simplification would be to use `operator()` as a generic way of calling functions.

However, this would probably add significant overhead during runtime.

And, adding additional complexity is the fact that with ChaiScript's flexibility it would be possible to add one (or more!) `operator()` overloads during the middle of a script execution, so there'd be no way to fix these things up front at parse time or during the first execution.

**Comment by klaasdijkstra** [comment]
Maybe just calling the "()" operator as real a function could be a nice workaround and still be efficient on runtime. See my example for the proposal:

The C++ class:

```
class Mat {
public:
	std::string operator()(const int i) {
		std::stringstream s;
		s << "I am (" << i << ")" << endl;
		return s.str();
	}
	std::string operator[](const int i) {
		std::stringstream s;
		s << "I am [" << i << "]" << endl;
		return s.str();
	}
};
```
Adding the function would work the same:

```
chai.add(chaiscript::user_type<Mat>(), "Mat");
chai.add(chaiscript::constructor<Mat()>(), "Mat");
chai.add(chaiscript::fun(static_cast<string(Mat::*)(const int)>(&Mat::operator())), "()");
chai.add(chaiscript::fun(static_cast<string(Mat::*)(const int)>(&Mat::operator[])), "[]");
```
However, in the current ChaiScript implementation only the "[]" works properly:
```
chai> var m = Mat()
chai> m[1]
I am [1]
chai> m(1)
Error: "'m' does not evaluate to a function." during evaluation at (1, 1)
```
My current, less elegant, workaround is to make a "one letter" function:

```
chai.add(chaiscript::fun(static_cast(&Mat::operator())), "f");
```

```
chai> m.f(1)
I am (1)
```
However, I propose to add the following syntax:
```
chai> m.(1)
Error: "Incomplete dot access fun call" during evaluation at (1, 3)
```
Currently this throws an error. But this would be a nice workaround.
